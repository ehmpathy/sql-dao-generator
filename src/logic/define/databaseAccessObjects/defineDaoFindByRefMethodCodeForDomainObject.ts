import { DomainObjectMetadata } from 'domain-objects-metadata';

import { defineOutputTypeOfFoundDomainObject } from './defineOutputTypeOfFoundDomainObject';

export const defineDaoFindByRefMethodCodeForDomainObject = ({
  domainObject,
}: {
  domainObject: DomainObjectMetadata;
}) => {
  // define the imports
  const imports = [
    ...new Set([
      // always present imports
      `import { UnexpectedCodePathError } from '@ehmpathy/error-fns';`,
      `import { Ref, isPrimaryKeyRef, isUniqueKeyRef } from 'domain-objects';`,
      `import { HasMetadata } from 'type-fns';`,

      '', // split module from relative imports
      `import { ${domainObject.name} } from '$PATH_TO_DOMAIN_OBJECT';`,
      "import { DatabaseConnection } from '$PATH_TO_DATABASE_CONNECTION';",
      `import { findByUnique } from './findByUnique';`,
      `import { findByUuid } from './findByUuid';`,
    ]),
  ];

  // define the output type
  const outputType = defineOutputTypeOfFoundDomainObject(domainObject);

  // define the content
  const code = `
${imports.join('\n')}

export const findByRef = async (
  input: { ref: Ref<typeof ${domainObject.name}> },
  context: { dbConnection: DatabaseConnection },
): Promise<${outputType} | null> => {
  if (isPrimaryKeyRef({ of: ${domainObject.name} })(input.ref))
    return await findByUuid(input.ref, context);
  if (isUniqueKeyRef({ of: ${domainObject.name} })(input.ref))
    return await findByUnique(input.ref, context);
  throw new UnexpectedCodePathError('invalid ref for ${
    domainObject.name
  }', { input });
};

`.trim();

  // return the code
  return code;
};
