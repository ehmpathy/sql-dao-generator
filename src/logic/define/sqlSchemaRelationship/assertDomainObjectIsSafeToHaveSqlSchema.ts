import { DomainObjectMetadata, DomainObjectVariant } from 'domain-objects-metadata';
import { UserInputError } from '../../UserInputError';

export const assertDomainObjectIsSafeToHaveSqlSchema = ({ domainObject }: { domainObject: DomainObjectMetadata }) => {
  // make sure that entities have "unique" and "updatable" defined
  if (domainObject.extends === DomainObjectVariant.DOMAIN_ENTITY) {
    if (!domainObject.decorations.unique?.length)
      throw new UserInputError({
        reason:
          "domain entities must have at least one 'unique' property defined in order for a schema to be generated.",
        domainObjectName: domainObject.name,
        potentialSolution:
          "(note: if its not unique on any natural keys, please specify the 'uuid' as it's only unique property. e.g., `['uuid']`) ",
      });
    if (!Array.isArray(domainObject.decorations.updatable))
      throw new UserInputError({
        reason:
          "domain entities must have their 'updatable' properties defined in order for a schema to be generated. ",
        domainObjectName: domainObject.name,
        potentialSolution:
          '(note: if it has no updatable properties, please specify this with an empty array. e.g., `[]`) ',
      });
  }

  // make sure that value objects do not have unique or updatable defined
  if (domainObject.extends === DomainObjectVariant.DOMAIN_VALUE_OBJECT) {
    if (domainObject.decorations.unique)
      throw new UserInputError({
        reason:
          "domain value objects must _not_ have their 'unique' properties specified. value objects are unique on all of their properties by definition.",
        domainObjectName: domainObject.name,
      });
    if (domainObject.decorations.updatable)
      throw new UserInputError({
        reason:
          "domain value objects must _not_ have any 'updatable' properties specified. value objects do not have any updatable properties by definition.",
        domainObjectName: domainObject.name,
      });
  }

  // make sure that events have "unique" defined
  if (domainObject.extends === DomainObjectVariant.DOMAIN_EVENT) {
    if (!domainObject.decorations.unique?.length)
      throw new UserInputError({
        reason: "domain events must have at least one 'unique' property defined in order for a schema to be generated",
        domainObjectName: domainObject.name,
      });
  }
};
