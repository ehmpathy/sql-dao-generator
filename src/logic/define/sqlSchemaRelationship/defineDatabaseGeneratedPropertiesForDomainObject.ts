import { DomainObjectMetadata } from 'domain-objects-metadata';
import { isPresent } from 'type-fns';

import { SqlSchemaPropertyMetadata } from '../../../domain/objects/SqlSchemaPropertyMetadata';

export const defineDatabaseGeneratedSqlSchemaPropertiesForDomainObject = ({
  domainObject,
}: {
  domainObject: DomainObjectMetadata;
}) => {
  const isUniqueOnUuid = domainObject.decorations.unique?.includes('uuid');
  const hasUpdatableProperties =
    (domainObject.decorations.updatable?.length ?? 0) > 0;

  // define which sql schema properties are generated by the db
  return [
    new SqlSchemaPropertyMetadata({
      name: 'id',
      isDatabaseGenerated: true,
      isArray: false,
      isNullable: false,
      isUpdatable: false,
      reference: null,
    }),
    !isUniqueOnUuid
      ? new SqlSchemaPropertyMetadata({
          name: 'uuid',
          isDatabaseGenerated: true,
          isArray: false,
          isNullable: false,
          isUpdatable: false,
          reference: null,
        })
      : null, // "uuid" is only autogenerated if the domain object is not unique on it
    new SqlSchemaPropertyMetadata({
      name: 'created_at',
      isDatabaseGenerated: true,
      isArray: false,
      isNullable: false,
      isUpdatable: false,
      reference: null,
    }),
    hasUpdatableProperties
      ? new SqlSchemaPropertyMetadata({
          name: 'effective_at',
          isDatabaseGenerated: true,
          isArray: false,
          isNullable: false,
          isUpdatable: false,
          reference: null,
        })
      : null, // domain objects only have "effective_at" defined if they're updatable
    hasUpdatableProperties
      ? new SqlSchemaPropertyMetadata({
          name: 'updated_at',
          isDatabaseGenerated: true,
          isArray: false,
          isNullable: false,
          isUpdatable: false,
          reference: null,
        })
      : null, // domain objects only have "updated_at" defined if they're updatable
  ].filter(isPresent);
};
