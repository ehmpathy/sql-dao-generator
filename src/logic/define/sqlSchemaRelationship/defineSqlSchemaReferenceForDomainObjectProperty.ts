// tslint:disable: max-classes-per-file
import { isPropertyNameAReferenceIntuitively } from 'domain-objects';
import {
  DomainObjectMetadata,
  DomainObjectPropertyMetadata,
  DomainObjectPropertyType,
  DomainObjectReferenceMetadata,
  DomainObjectVariant,
  isDomainObjectArrayProperty,
  isDomainObjectReferenceProperty,
} from 'domain-objects-metadata';
import omit from 'lodash.omit';

import {
  SqlSchemaReferenceMetadata,
  SqlSchemaReferenceMethod,
} from '../../../domain/objects/SqlSchemaReferenceMetadata';
import { UnexpectedCodePathDetectedError } from '../../UnexpectedCodePathDetectedError';
import { UserInputError } from '../../UserInputError';
import { getDomainObjectNameThatPropertyIsUnambigiouslyNaturallyNamedAfter } from './getDomainObjectNameThatPropertyIsUnambigiouslyNaturallyNamedAfter/getDomainObjectNameThatPropertyIsUnambigiouslyNaturallyNamedAfter';

export class PropertyReferencingDomainObjectNotNamedCorrectlyError extends Error {
  constructor({
    domainObject,
    property,
    referencedDomainObject,
  }: {
    domainObject: DomainObjectMetadata;
    property: DomainObjectPropertyMetadata;
    referencedDomainObject: {
      expected: string;
      inferred: string | null;
    };
  }) {
    super(
      `
Properties that reference a domain-object must be named after the domain-object they reference. '${domainObject.name}.${property.name}: ${referencedDomainObject.expected}' does not meet this criteria.

For example:
- allowed:
  - 'address: Address'
  - 'homeAddress: Address'
  - 'address: HomeAddress'
  - 'homeAddress: HomeAddress'
  - 'home: HomeAddress'
- not allowed:
  - 'home: Address'

Ambiguity:
- expected: ${referencedDomainObject.expected}
- inferred: ${referencedDomainObject.inferred}

Tip: the prefixes or suffixes should unambiguously match
`.trim(),
    );
  }
}

export class DirectlyNestedNonDomainObjectReferenceForbiddenError extends Error {
  constructor({
    domainObject,
    property,
    referencedDomainObject,
  }: {
    domainObject: DomainObjectMetadata;
    property: DomainObjectPropertyMetadata;
    referencedDomainObject: DomainObjectReferenceMetadata;
  }) {
    super(
      `
${referencedDomainObject.extends} found directly nested inside of another domain object. '${domainObject.name}.${property.name}' reference '${referencedDomainObject.name}'

This is not allowed, as this is bad practice when persisting domain-objects due to maintainability problems with this pattern in backend code. See the readme for more details.

Instead, reference the entity by 'uuid' in the backend. For example:

\`\`\`ts
-- say you're referencing this domain-entity
export class User extends DomainEntity<User> implements User {};

-- instead of this
interface Profile {
  user: User; -- forbidden: referencing domain entity by nesting
  ...
}

-- do this
interface Profile {
  userUuid: string; -- suggested: reference it by uuid
  ...
}
\`\`\`

Note: the generated sql-schema will be the same as if it was a nested reference, but the domain-object referencing by uuid will make your backend code easier to maintain.
  `.trim(),
    );
  }
}

export const defineSqlSchemaReferenceForDomainObjectProperty = ({
  property,
  domainObject,
  allDomainObjects,
}: {
  property: DomainObjectPropertyMetadata;
  domainObject: DomainObjectMetadata;
  allDomainObjects: DomainObjectMetadata[];
}): SqlSchemaReferenceMetadata | null => {
  // determine what kind of reference it can be
  const isDirectNestedReferenceCandidate =
    isDomainObjectReferenceProperty(property);
  const isDirectNestedReferenceArrayCandidate =
    isDomainObjectArrayProperty(property) &&
    isDomainObjectReferenceProperty(property.of);
  const isImplicitUuidReferenceCandidate =
    property.type === DomainObjectPropertyType.STRING &&
    new RegExp(/Uuid/).test(property.name);
  const isImplicitUuidReferenceArrayCandidate =
    isDomainObjectArrayProperty(property) &&
    property.of.type === DomainObjectPropertyType.STRING &&
    new RegExp(/Uuids/).test(property.name);

  // handle direct nested references
  if (
    isDirectNestedReferenceCandidate ||
    isDirectNestedReferenceArrayCandidate
  ) {
    // grab the referenced object
    const referencedDomainObject = (() => {
      if (isDirectNestedReferenceCandidate)
        return property.of as DomainObjectReferenceMetadata;
      if (isDirectNestedReferenceArrayCandidate)
        return (property.of as DomainObjectPropertyMetadata)
          .of as DomainObjectReferenceMetadata;
      throw new UnexpectedCodePathDetectedError({
        reason:
          'should have met one of the above criteria for sql schema reference definition',
        domainObjectName: domainObject.name,
        domainObjectPropertyName: property.name,
      }); // fail fast
    })();

    // check that the property-name is unambiguously and naturally named after this domain object
    const domainObjectNamePropertyIsNamedAfter =
      getDomainObjectNameThatPropertyIsUnambigiouslyNaturallyNamedAfter({
        parentDomainObjectName: domainObject.name,
        propertyName: property.name,
        allDomainObjectNames: [
          ...new Set([
            ...allDomainObjects.map(({ name }) => name),
            referencedDomainObject.name,
          ]),
        ], // note: We add the referenced domain object name her because we _have_ the reference, even if its not in the "metadatas" of available domain objects; this avoids a class of errors from technically being possible (ones that _shouldn't occur in real life)
      });
    if (domainObjectNamePropertyIsNamedAfter !== referencedDomainObject.name)
      throw new PropertyReferencingDomainObjectNotNamedCorrectlyError({
        property,
        domainObject,
        referencedDomainObject: {
          expected: referencedDomainObject.name,
          inferred: domainObjectNamePropertyIsNamedAfter,
        },
      });

    // check that the domain object referenced by direct nesting is not a domain entity or a domain event
    if (referencedDomainObject.extends !== DomainObjectVariant.DOMAIN_LITERAL)
      throw new DirectlyNestedNonDomainObjectReferenceForbiddenError({
        domainObject,
        property,
        referencedDomainObject,
      });

    // if the above passes, then we're good to move forward
    return new SqlSchemaReferenceMetadata({
      method: SqlSchemaReferenceMethod.DIRECT_BY_NESTING,
      of: new DomainObjectReferenceMetadata(referencedDomainObject),
    });
  }

  // handle inferred uuid references
  const referenceOfUuid = (() => {
    if (
      !isImplicitUuidReferenceCandidate &&
      !isImplicitUuidReferenceArrayCandidate
    )
      return null; // no reference if not a candidate of either
    const foundReferencedDomainObjectName =
      getDomainObjectNameThatPropertyIsUnambigiouslyNaturallyNamedAfter({
        parentDomainObjectName: domainObject.name,
        propertyName: property.name,
        allDomainObjectNames: allDomainObjects.map(({ name }) => name),
      });
    const foundReferencedDomainObjectMetadata = allDomainObjects.find(
      (domainObject) => domainObject.name === foundReferencedDomainObjectName,
    );
    if (!foundReferencedDomainObjectMetadata) return null;
    if (
      foundReferencedDomainObjectMetadata.extends ===
      DomainObjectVariant.DOMAIN_LITERAL
    )
      // safety check
      throw new UserInputError({
        reason:
          'domain-literals persisted in the same database as this domain-object should not be referenced by implicit uuid reference. reference them by direct nesting instead.',
        domainObjectName: domainObject.name,
        domainObjectPropertyName: property.name,
      });
    return foundReferencedDomainObjectMetadata;
  })();
  if (referenceOfUuid)
    return new SqlSchemaReferenceMetadata({
      method: SqlSchemaReferenceMethod.IMPLICIT_BY_UUID,
      of: new DomainObjectReferenceMetadata(
        omit(referenceOfUuid, ['properties', 'decorations']),
      ),
    });

  // otherwise, no reference
  return null;
};
