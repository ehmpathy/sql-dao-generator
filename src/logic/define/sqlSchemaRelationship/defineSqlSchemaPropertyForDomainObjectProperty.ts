import { camelCase, snakeCase } from 'change-case';
import { DomainObjectMetadata, DomainObjectPropertyMetadata, DomainObjectPropertyType } from 'domain-objects-metadata';

import { SqlSchemaPropertyMetadata } from '../../../domain/objects/SqlSchemaPropertyMetadata';
import { SqlSchemaReferenceMethod } from '../../../domain/objects/SqlSchemaReferenceMetadata';
import { UnexpectedCodePathDetectedError } from '../../UnexpectedCodePathDetectedError';
import { UserInputError } from '../../UserInputError';
import { defineDatabaseGeneratedSqlSchemaPropertiesForDomainObject } from './defineDatabaseGeneratedPropertiesForDomainObject';
import { defineSqlSchemaReferenceForDomainObjectProperty } from './defineSqlSchemaReferenceForDomainObjectProperty';

export const defineSqlSchemaPropertyForDomainObjectProperty = ({
  property,
  domainObject,
  allDomainObjects,
}: {
  property: DomainObjectPropertyMetadata;
  domainObject: DomainObjectMetadata;
  allDomainObjects: DomainObjectMetadata[];
}): SqlSchemaPropertyMetadata => {
  // sanity check that property name is in camel case
  if (camelCase(property.name) !== property.name)
    throw new UserInputError({
      reason: 'property names must be camel case',
      domainObjectName: domainObject.name,
      domainObjectPropertyName: property.name,
    });

  // sanity check that this is not called for a database generated property (since that should be handled elsewhere)
  const databaseGeneratedProperties = defineDatabaseGeneratedSqlSchemaPropertiesForDomainObject({ domainObject });
  if (databaseGeneratedProperties.map(({ name }) => name).includes(snakeCase(property.name)))
    throw new UnexpectedCodePathDetectedError({
      reason: '`defineSqlSchemaPropertyForDomainObjectProperty was called for an db-generated property',
      domainObjectName: domainObject.name,
      domainObjectPropertyName: property.name,
    });

  // define data about this property
  const isUpdatable = domainObject.decorations.updatable?.includes(property.name) ?? false;
  const isNullable = !!property.nullable;
  const isArray = property.type === DomainObjectPropertyType.ARRAY;

  // determine if this is a reference of another domain object
  const reference = defineSqlSchemaReferenceForDomainObjectProperty({
    property,
    domainObject,
    allDomainObjects,
  });

  // define the schema property name
  const schemaPropertyName = (() => {
    const baseSchemaPropertyName = snakeCase(property.name);
    if (reference) {
      if (isArray) {
        if (reference.method === SqlSchemaReferenceMethod.DIRECT_BY_NESTING)
          return `${baseSchemaPropertyName.replace(/s$/, '')}_ids`; // suffix w/ ids, since its an array of fks
        if (reference.method === SqlSchemaReferenceMethod.IMPLICIT_BY_UUID)
          return `${baseSchemaPropertyName.replace(/_uuids$/, '')}_ids`; // suffix w/ ids, since its an array of fks
      }
      if (reference.method === SqlSchemaReferenceMethod.DIRECT_BY_NESTING) return `${baseSchemaPropertyName}_id`; // suffix w/ id, since its a fk
      if (reference.method === SqlSchemaReferenceMethod.IMPLICIT_BY_UUID)
        return `${baseSchemaPropertyName.replace(/_uuid$/, '')}_id`; // suffix w/ id, since its a fk
    }
    return baseSchemaPropertyName; // otherwise, its just the regular name, no modifications
  })();

  return new SqlSchemaPropertyMetadata({
    name: schemaPropertyName,
    isUpdatable,
    isNullable,
    isArray,
    isDatabaseGenerated: false, // this method is only called for properties that are not database generated (as checked above)
    reference,
  });
};
