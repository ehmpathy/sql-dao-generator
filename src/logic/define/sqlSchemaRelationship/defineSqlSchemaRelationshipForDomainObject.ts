import { snakeCase } from 'change-case';
import {
  DomainObjectMetadata,
  DomainObjectPropertyType,
  DomainObjectVariant,
} from 'domain-objects-metadata';

import { SqlSchemaToDomainObjectRelationship } from '../../../domain/objects/SqlSchemaToDomainObjectRelationship';
import { UnexpectedCodePathDetectedError } from '../../UnexpectedCodePathDetectedError';
import { UserInputError } from '../../UserInputError';
import { assertDomainObjectIsSafeToHaveSqlSchema } from './assertDomainObjectIsSafeToHaveSqlSchema';
import { defineDatabaseGeneratedSqlSchemaPropertiesForDomainObject } from './defineDatabaseGeneratedPropertiesForDomainObject';
import { defineSqlSchemaPropertyForDomainObjectProperty } from './defineSqlSchemaPropertyForDomainObjectProperty';
import { isNotADatabaseGeneratedProperty } from './isNotADatabaseGeneratedProperty';

export const defineSqlSchemaRelationshipForDomainObject = ({
  domainObject,
  allDomainObjects,
}: {
  domainObject: DomainObjectMetadata;
  allDomainObjects: DomainObjectMetadata[];
}): SqlSchemaToDomainObjectRelationship => {
  // figure out some relevant info
  const sqlSchemaName = snakeCase(domainObject.name); // names are in snake case

  // define the properties
  const databaseGeneratedSqlSchemaProperties =
    defineDatabaseGeneratedSqlSchemaPropertiesForDomainObject({
      domainObject,
    });
  const propertiesRelationship = [
    /// autogenerated values
    ...databaseGeneratedSqlSchemaProperties.map((sqlSchemaProperty) => {
      // see if this property was defined on the domain object (i.e., they want the value for it)
      const domainObjectProperty =
        Object.values(domainObject.properties).find(
          (property) => snakeCase(property.name) === sqlSchemaProperty.name,
        ) ?? null;

      // if so, validate that it has the correct type
      if (domainObjectProperty) {
        const expectedType = (() => {
          if (sqlSchemaProperty.name === 'id')
            return DomainObjectPropertyType.NUMBER;
          if (sqlSchemaProperty.name === 'uuid')
            return DomainObjectPropertyType.STRING;
          if (sqlSchemaProperty.name === 'created_at')
            return DomainObjectPropertyType.DATE;
          if (sqlSchemaProperty.name === 'effective_at')
            return DomainObjectPropertyType.DATE;
          if (sqlSchemaProperty.name === 'updated_at')
            return DomainObjectPropertyType.DATE;
          throw new UnexpectedCodePathDetectedError({
            reason:
              'unexpected database generated property name. this is a bug within sql-schema-generator.', // fail fast if expectations not met
            domainObjectName: domainObject.name,
            domainObjectPropertyName: sqlSchemaProperty.name,
          });
        })();
        if (domainObjectProperty.type !== expectedType)
          throw new UserInputError({
            reason: `reserved database-generated property does not match expected type. expected ${expectedType} but received ${domainObjectProperty.type}`,
            domainObjectName: domainObject.name,
            domainObjectPropertyName: domainObjectProperty.name,
          });
      }

      // return the relationship
      return {
        sqlSchema: sqlSchemaProperty,
        domainObject: domainObjectProperty,
      };
    }),

    // non autogenerated values
    ...Object.values(domainObject.properties)
      .filter(
        (domainObjectProperty) =>
          !databaseGeneratedSqlSchemaProperties
            .map(({ name }) => name)
            .includes(snakeCase(domainObjectProperty.name)), // filter out the db generated values, because we already got them above
      )
      .map((domainObjectProperty) => {
        return {
          sqlSchema: defineSqlSchemaPropertyForDomainObjectProperty({
            property: domainObjectProperty,
            domainObject,
            allDomainObjects,
          }),
          domainObject: domainObjectProperty,
        };
      }),
  ];

  // define the unique keys
  const sqlSchemaUniqueKeys = (() => {
    // define what the unique prop names are, based on what kind of object it is
    const domainObjectPropertyNamesInUniqueKey =
      domainObject.extends === DomainObjectVariant.DOMAIN_LITERAL
        ? propertiesRelationship // if its a literal, the schema is unique on all properties of the object (excluding autogenerated ones like id and uuid)
            .filter(isNotADatabaseGeneratedProperty)
            .map(
              ({ domainObject: domainObjectProperty }) =>
                domainObjectProperty.name,
            )
        : domainObject.decorations.unique; // otherwise, its unique on what was specified

    // convert the domain object prop names into sql schema prop names
    return (
      domainObjectPropertyNamesInUniqueKey?.map((domainObjectPropertyName) => {
        const sqlSchemaPropertyName = propertiesRelationship.find(
          (propertyRelationship) =>
            propertyRelationship.domainObject?.name ===
            domainObjectPropertyName,
        )?.sqlSchema.name;
        if (!sqlSchemaPropertyName)
          throw new UserInputError({
            reason: 'Unique keys must be properties of the domain object.',
            domainObjectName: domainObject.name,
            potentialSolution: ` Was there a typo defining '${
              domainObject.name
            }.unique' with key name '${domainObjectPropertyName}'? (Is it one of ${JSON.stringify(
              Object.keys(domainObject.properties),
            )}?)`,
          });
        return sqlSchemaPropertyName;
      }) ?? null
    );
  })();

  // make sure that this domain object is safe to have a schema defined for it
  assertDomainObjectIsSafeToHaveSqlSchema({ domainObject });

  // return the relationship
  return new SqlSchemaToDomainObjectRelationship({
    name: {
      sqlSchema: sqlSchemaName,
      domainObject: domainObject.name,
    },
    properties: propertiesRelationship,
    decorations: {
      alias: {
        domainObject: domainObject.decorations.alias,
      },
      unique: {
        sqlSchema: sqlSchemaUniqueKeys,
        domainObject: domainObject.decorations.unique,
      },
    },
  });
};
