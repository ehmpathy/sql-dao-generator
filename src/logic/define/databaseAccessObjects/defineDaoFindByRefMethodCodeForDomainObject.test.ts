import {
  DomainObjectMetadata,
  DomainObjectPropertyType,
  DomainObjectVariant,
} from 'domain-objects-metadata';

import { defineSqlSchemaRelationshipForDomainObject } from '../sqlSchemaRelationship/defineSqlSchemaRelationshipForDomainObject';
import { defineDaoFindByRefMethodCodeForDomainObject } from './defineDaoFindByRefMethodCodeForDomainObject';

describe('defineDaoFindByRefMethodCodeForDomainObject', () => {
  it('should look correct for simple literal', () => {
    // define what we're testing on
    const domainObject = new DomainObjectMetadata({
      name: 'Geocode',
      extends: DomainObjectVariant.DOMAIN_LITERAL,
      properties: {
        id: { name: 'id', type: DomainObjectPropertyType.NUMBER },
        uuid: { name: 'uuid', type: DomainObjectPropertyType.STRING },
        latitude: { name: 'latitude', type: DomainObjectPropertyType.NUMBER },
        longitude: {
          name: 'longitude',
          type: DomainObjectPropertyType.NUMBER,
        },
      },
      decorations: {
        alias: null,
        primary: null,
        unique: null,
        updatable: null,
      },
    });

    // run it
    const code = defineDaoFindByRefMethodCodeForDomainObject({
      domainObject,
    });

    // log an example
    expect(code).toContain('input: { ref: Ref<typeof Geocode> }');
    expect(code).toMatchSnapshot();
  });
  it('should look correct for simple domain entity', () => {
    // define what we're testing on
    const domainObject = new DomainObjectMetadata({
      name: 'Carriage',
      extends: DomainObjectVariant.DOMAIN_ENTITY,
      properties: {
        id: {
          name: 'id',
          type: DomainObjectPropertyType.NUMBER,
          required: false,
        },
        uuid: {
          name: 'uuid',
          type: DomainObjectPropertyType.STRING,
          required: false,
        },
        cin: {
          name: 'cin',
          type: DomainObjectPropertyType.STRING,
          required: true,
        },
        carries: {
          name: 'carries',
          type: DomainObjectPropertyType.ENUM,
          of: ['PASSENGER', 'FREIGHT'],
          required: true,
        },
        capacity: {
          name: 'capacity',
          type: DomainObjectPropertyType.NUMBER,
          nullable: true,
        },
      },
      decorations: {
        alias: null,
        primary: null,
        unique: ['cin'],
        updatable: ['capacity'],
      },
    });

    // run it
    const code = defineDaoFindByRefMethodCodeForDomainObject({
      domainObject,
    });

    // log an example
    expect(code).toContain('input: { ref: Ref<typeof Carriage> }');
    expect(code).toMatchSnapshot();
  });
  it('should look correct for a domain event with a static referenced array', () => {
    // define what we're testing on
    const domainObject = new DomainObjectMetadata({
      name: 'TrainLocatedEvent',
      extends: DomainObjectVariant.DOMAIN_EVENT,
      properties: {
        id: {
          name: 'id',
          type: DomainObjectPropertyType.NUMBER,
          required: false,
        },
        trainUuid: {
          name: 'trainUuid',
          type: DomainObjectPropertyType.STRING,
          required: true,
        },
        occurredAt: {
          name: 'occurredAt',
          type: DomainObjectPropertyType.DATE,
          required: true,
        },
        geocodes: {
          name: 'geocodes',
          type: DomainObjectPropertyType.ARRAY,
          of: {
            type: DomainObjectPropertyType.REFERENCE,
            of: {
              name: 'Geocode',
              extends: DomainObjectVariant.DOMAIN_LITERAL,
            },
          },
          required: true,
        },
      },
      decorations: {
        alias: null,
        primary: null,
        unique: ['trainUuid', 'occurredAt'],
        updatable: [],
      },
    });

    // run it
    const code = defineDaoFindByRefMethodCodeForDomainObject({
      domainObject,
    });

    // log an example
    expect(code).toContain('input: { ref: Ref<typeof TrainLocatedEvent> }');
    expect(code).toMatchSnapshot();
  });
  it('should look correct for domain entity with references, array and solo, implicit and direct', () => {
    // define what we're testing on
    const domainObject = new DomainObjectMetadata({
      name: 'Train',
      extends: DomainObjectVariant.DOMAIN_ENTITY,
      properties: {
        id: {
          name: 'id',
          type: DomainObjectPropertyType.NUMBER,
          required: false,
        },
        uuid: {
          name: 'uuid',
          type: DomainObjectPropertyType.STRING,
          required: false,
        },
        tin: {
          name: 'tin',
          type: DomainObjectPropertyType.STRING,
          required: true,
        },
        homeStationGeocode: {
          name: 'homeStationGeocode',
          type: DomainObjectPropertyType.REFERENCE,
          of: {
            name: 'Geocode',
            extends: DomainObjectVariant.DOMAIN_LITERAL,
          },
        },
        leadEngineerUuid: {
          name: 'leadEngineerUuid',
          type: DomainObjectPropertyType.STRING,
        },
        badges: {
          name: 'badges',
          type: DomainObjectPropertyType.ARRAY,
          of: {
            type: DomainObjectPropertyType.REFERENCE,
            of: {
              name: 'TrainBadge',
              extends: DomainObjectVariant.DOMAIN_LITERAL,
            },
          },
        },
        locomotiveUuids: {
          name: 'locomotiveUuids',
          type: DomainObjectPropertyType.ARRAY,
          of: {
            type: DomainObjectPropertyType.STRING,
          },
        },
      },
      decorations: {
        alias: null,
        primary: null,
        unique: ['tin'],
        updatable: [
          'homeStation',
          'badges',
          'locomotiveUuids',
          'leadEngineerUuid',
        ],
      },
    });

    // run it
    const code = defineDaoFindByRefMethodCodeForDomainObject({
      domainObject,
    });

    // log an example
    expect(code).toContain('isPrimaryKeyRef({ of: Train })');
    expect(code).toMatchSnapshot();
  });
});
