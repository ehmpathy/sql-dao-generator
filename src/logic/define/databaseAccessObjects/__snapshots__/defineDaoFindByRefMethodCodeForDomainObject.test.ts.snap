// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`defineDaoFindByRefMethodCodeForDomainObject should look correct for a domain event with a static referenced array 1`] = `
"import { UnexpectedCodePathError } from '@ehmpathy/error-fns';
import { Ref, isPrimaryKeyRef, isUniqueKeyRef } from 'domain-objects';
import { HasMetadata } from 'type-fns';

import { TrainLocatedEvent } from '$PATH_TO_DOMAIN_OBJECT';
import { DatabaseConnection } from '$PATH_TO_DATABASE_CONNECTION';
import { findByUnique } from './findByUnique';
import { findByUuid } from './findByUuid';

export const findByRef = async (
  input: { ref: Ref<typeof TrainLocatedEvent> },
  context: { dbConnection: DatabaseConnection },
): Promise<HasMetadata<TrainLocatedEvent> | null> => {
  if (isPrimaryKeyRef({ of: TrainLocatedEvent })(input.ref))
    return await findByUuid(input.ref, context);
  if (isUniqueKeyRef({ of: TrainLocatedEvent })(input.ref))
    return await findByUnique(input.ref, context);
  throw new UnexpectedCodePathError('invalid ref for TrainLocatedEvent', { input });
};"
`;

exports[`defineDaoFindByRefMethodCodeForDomainObject should look correct for domain entity with references, array and solo, implicit and direct 1`] = `
"import { UnexpectedCodePathError } from '@ehmpathy/error-fns';
import { Ref, isPrimaryKeyRef, isUniqueKeyRef } from 'domain-objects';
import { HasMetadata } from 'type-fns';

import { Train } from '$PATH_TO_DOMAIN_OBJECT';
import { DatabaseConnection } from '$PATH_TO_DATABASE_CONNECTION';
import { findByUnique } from './findByUnique';
import { findByUuid } from './findByUuid';

export const findByRef = async (
  input: { ref: Ref<typeof Train> },
  context: { dbConnection: DatabaseConnection },
): Promise<HasMetadata<Train> | null> => {
  if (isPrimaryKeyRef({ of: Train })(input.ref))
    return await findByUuid(input.ref, context);
  if (isUniqueKeyRef({ of: Train })(input.ref))
    return await findByUnique(input.ref, context);
  throw new UnexpectedCodePathError('invalid ref for Train', { input });
};"
`;

exports[`defineDaoFindByRefMethodCodeForDomainObject should look correct for simple domain entity 1`] = `
"import { UnexpectedCodePathError } from '@ehmpathy/error-fns';
import { Ref, isPrimaryKeyRef, isUniqueKeyRef } from 'domain-objects';
import { HasMetadata } from 'type-fns';

import { Carriage } from '$PATH_TO_DOMAIN_OBJECT';
import { DatabaseConnection } from '$PATH_TO_DATABASE_CONNECTION';
import { findByUnique } from './findByUnique';
import { findByUuid } from './findByUuid';

export const findByRef = async (
  input: { ref: Ref<typeof Carriage> },
  context: { dbConnection: DatabaseConnection },
): Promise<HasMetadata<Carriage> | null> => {
  if (isPrimaryKeyRef({ of: Carriage })(input.ref))
    return await findByUuid(input.ref, context);
  if (isUniqueKeyRef({ of: Carriage })(input.ref))
    return await findByUnique(input.ref, context);
  throw new UnexpectedCodePathError('invalid ref for Carriage', { input });
};"
`;

exports[`defineDaoFindByRefMethodCodeForDomainObject should look correct for simple literal 1`] = `
"import { UnexpectedCodePathError } from '@ehmpathy/error-fns';
import { Ref, isPrimaryKeyRef, isUniqueKeyRef } from 'domain-objects';
import { HasMetadata } from 'type-fns';

import { Geocode } from '$PATH_TO_DOMAIN_OBJECT';
import { DatabaseConnection } from '$PATH_TO_DATABASE_CONNECTION';
import { findByUnique } from './findByUnique';
import { findByUuid } from './findByUuid';

export const findByRef = async (
  input: { ref: Ref<typeof Geocode> },
  context: { dbConnection: DatabaseConnection },
): Promise<HasMetadata<Geocode> | null> => {
  if (isPrimaryKeyRef({ of: Geocode })(input.ref))
    return await findByUuid(input.ref, context);
  if (isUniqueKeyRef({ of: Geocode })(input.ref))
    return await findByUnique(input.ref, context);
  throw new UnexpectedCodePathError('invalid ref for Geocode', { input });
};"
`;
